### PR TITLE
Rough sketch of an XChaCha security proof, derived from the XSalsa20 Paper

### PR DESCRIPTION
**Pull request by paragonie-scott**

### CODE REVIEW

File: xchacha.md [hunk 1]
```diff
@@ -213,6 +213,76 @@ xchacha20_encrypt(key, nonce, plaintext):
...
+2. Q(i) is a public computation of uniform random strings from uniform random
+   strings.
+
+Thus, HSalsa20 uses the same indices as the public inputs (constant + nonce)
+for its final output.
+
+The reliance on public computation for the security proof makes sense, and
+can be applied to ChaCha with a slight tweak.
+
+ChaCha is a slightly different than Salsa20: The constants occupy the indices
```
Comment (ValarDragon) [inline xchacha.md]: should be `is a slightly different construction than`

File: xchacha.md [hunk 2]
```diff
@@ -213,6 +213,76 @@ xchacha20_encrypt(key, nonce, plaintext):
...
+for its final output.
+
+The reliance on public computation for the security proof makes sense, and
+can be applied to ChaCha with a slight tweak.
+
+ChaCha is a slightly different construction than Salsa20: The constants occupy
+the indices at 0, 1, 2, 3. Meanwhile, the nonce populates indices 12, 13, 14, 15.
+ 
+Consequently, we can extract a public computation from ChaCha20 by selecting
+these indices from HChaCha20's final state, and the same security proof can
```
Comment (ValarDragon) [inline xchacha.md]: This should be 

> by selecting these indices _as_ HChaCha20's final state

right?

Comment (paragonie-scott) [reply xchacha.md]: It's selecting indices from the cipher's state after 20 rounds to generate the output.

Comment (paragonie-scott) [reply xchacha.md]: I just pushed another commit which aims to make this intent clearer. Does that help?

Comment (ValarDragon) [reply xchacha.md]: Thanks for the clarification, that makes sense

**Comment by ValarDragon** [comment]
This looks good to me. I think it would be valuable to note the security assumption of HSalsa somewhere as well, since thats the comparison being made. (I.e. its secure as long as salsa20 is secure)

**Pull request merged**